src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -390,6 +464,29 @@ async def run_setup_py(
     return RunSetupPyResult(output_digest)
 
 
+@rule
+async def determine_setup_kwargs(
+    exported_target: ExportedTarget, union_membership: UnionMembership
+) -> SetupKwargs:
+    target = exported_target.target
+    plugin_requests = union_membership.get(SetupKwargsPluginRequest)  # type: ignore[misc]
+
+    # If no plugins, simply return what the user explicitly specified in a BUILD file.
+    if not plugin_requests:
+        return SetupKwargs(exported_target.provides.kwargs, address=target.address)
+
+    if len(plugin_requests) > 1:
```

Comment:
This feels like a clunky use of union. I would have thought that a more idiomatic way of doing this is to require a custom target type and a @rule that maps it to an appropriate union member?

Reply:
This is how we use unions for basically everything now, including `FieldSet`. For example, a `FieldSet` calls `FieldSet.is_valid()`.

Sure, it looks clunky, but also remember that this clunkiness is isolated to this particular rule definition.

> require a custom target type

Why require that? For example, in the Pants repo, we always want to use our custom implementation. It would be annoying to have to add a custom target type and it would be easy for us to forgot to use it.

Instead, in https://www.pantsbuild.org/v2.0/docs/plugins-setup-py-goal, I'm using the language "Consider defining a custom `python_distribution` target type". It could make sense if they only want to sometimes use their implementation, but it's not always necessary.

Reply:
This use of union doesn't seem to require a union at all. You could simply get all the `__subclasses__` from a base class instead, no?

Reply:
Indeed. In fact, John wants to get rid of `UnionRule` and always look at `__subclassess__` for union bases. For 95% of our unions, that would be possible. The only reason we can't is `PluginField`.

(And, I'm skeptical of looking at `__subclasses__` because it means that you wouldn't be able to have a helper superclass anymore; that would inadvertently get registered as a union member, even though it's not intended to be.)

In the meantime, this is an idiomatic usage of unions. Unions changed quite a bit when we added the Target API, as it was no longer possible to do something like `await Get(TestResult, TestTarget, my_python_tests_target)` because this wouldn't handle subclasses of `PythonTests` properly. Again, almost every single union call site looks like this here, e.g. https://github.com/pantsbuild/pants/pull/9856 adding sugar back in May to better support this idiom.

Reply:
Stepping back a sec, AFAIK unions exist to provide the following functionality, supported by the engine: A union member instance can satisfy a request for a union type param. That is, `await Get(ResultType, UnionType, union_member_instance)` works. Why are we using them for not-this in the first place? Why not just use a simpler mechanism (`__subclasses__` being an obvious example).

Reply:
@stuhood would love to hear your thoughts as well.

Reply:
> That is, await Get(ResultType, UnionType, union_member_instance) works

To clarify, we are still doing that here. We call `await Get(SetupKwargs, SetupKwargsPluginRequest, my_pants_setup_kwargs_request)`. No matter what, we need to use the Engine's mechanism of unions.

What I'm discussing with `__subclasses__` is that we could get rid of the _registration_ of the union relationship so that it's more automatic. Rather than making the user explicitly express the relationship via `UnionRule(SetupKwargsPluginRequest, MyRequest)`, we could be clever and look up `__subclasses__` to find the all bases, and then automatically register those relationships with the engine.

The code in this rule would still be identical. All it would change is that one line `UnionRule(SetupKwargsPluginRequest, MyRequest)` in the user's plugin `register.py` file.

Reply:
This still seems like a lot of hoop-jumping to dynamically decide which union member to apply, when it seems that we could just let the engine figure it out. E.g., we could make a `PythonArtifactBase` union and call `await Get(SetupKwargs, PythonArtifactBase, target.provides)`. And have the routing chosen by the type of `target.provides`.
We provide the default `PythonArtifact` via the `setup_py` CAOF, but users could provide their own.  What are the downsides of that?

Reply:
> E.g., we could make a PythonArtifactBase union and call await Get(SetupKwargs, PythonArtifactBase, target.provides)

For this particular implementation, this breaks the Target API because `target.provides` is a `PythonProvidesField`. We encourage subclassing of `Field`s, and that needs to continue working, i.e. that you can subclass a `Field` without breaking any functionality.

But, I think that your suggestion is actually about subclassing `PythonArtifact`, right? Sure, we could do that. That was one of the original designs I considered. But I think it has several downsides compared to this approach:

1. Users must subclass `PythonArtifact`, including registering it in the legacy `BuildFileAliases` type (which we want to move away from). They also must remember to use that subclass in their BUILD files, or they'll accidentally not get their plugin behavior.
2. There's less flexibility for plugin authors to decide when their implementation should be used. For example, a plugin author may _always_ want their implementation to be used, like we do in this repo—so they can return `True` to `is_valid()`. Or, a plugin author might want to do something like inspect `target.address` so that all addresses in folder `project1/` use plugin1, and all addresses in `project2/` use `plugin2`. Or, they might want it to be based on the identity of the `PythonArtificat`, which they can still do via `is_valid()`. We let the plugin author decide what makes sense for them, rather than requiring that they create a subclass and BUILD file authors remember to use the correct subclass.

I'm a little confused, tbh, what the concern is here. This is a 23-line rule that plugin authors will likely never interact with. It's also using a pattern that we use in several other places, including codegen, all `FieldSet`s like `lint` and `test`, and dep inference:

https://github.com/pantsbuild/pants/blob/6b794475ec3e7d6d47f2446d87440d3c822240a2/src/python/pants/engine/internals/graph.py#L590-L601

Reply:
The concern is just that this is a clunky idiom. We have a streamlined way of "switching" on the runtime subtype of a param, and instead of leaning into it (and, yes, requiring users to use the correct subclass in BUILD files) we're using `is_valid`. Granted, `is_valid` provides more freedom, as you mentioned. But is that needed? Explaining `is_valid` is not simple. 

Anyway, I won't push on this.  But for now, at least, let's sort out the naming: "entry point" has specific meaning (the main method of a binary) so let's not use that terminology. Ditto, "plugin" means something specific in Pants, so let's not abuse that term. It may be true that this kwargs "plugin" lives in a plugin, but that's not a strict enough association to justify the terminology duplication. And as you say, you might have more than one. And finally, maybe we can rename `is_valid` to `is_applicable`, since we're not actually validating anything.

Reply:
> (and, yes, requiring users to use the correct subclass in BUILD files) 

We've generally accepted that we're okay with a little clunkiness in our code for a much better UX. I argue that applies here too. Sure, this 23 line rule might be clunky, but I think it's worth it due to the benefits it brings users, like allowing a user to not have to remember to use the subclass in a BUILD file.

> maybe we can rename is_valid to is_applicable

I like it. I'll fix `FieldSet` too.